open_esplibs/libmain/ets_timer.c

```diff
@@ -0,0 +1,391 @@
+ *
+ * | Timer:      | T0 | T1 | T2 | T3 |
+ * |-------------|----|----|----|----|
+ * | fire_ticks: | 10 | 20 | 30 | 40 |
+ * | next:       | T1 | T2 | T3 | 0  |
+ *
+ *
+ * For example we need to add a timer that should fire at 25 ticks:
+ *
+ * | Timer:      | T0 | T1  | new | T2 | T3 |
+ * |-------------|----|-----|-----|----|----|
+ * | fire_ticks: | 10 | 20  | 25  | 30 | 40 |
+ * | next:       | T1 | new | T2  | T3 | 0  |
+ *
+ * We squeeze the timer into the list so the list will always remain sorted
+ *
+ * Note: if add the same timer twice the system halts
+*/
+static void add_pending_timer(uint32_t ticks, ets_timer_t *timer)
+{
+    ets_timer_t *prev = 0;
+    ets_timer_t *curr = timer_list;
+    while (curr) {
+        if (((int32_t)ticks - (int32_t)curr->fire_ticks) < 1) {
+            // found a timer that should fire later
+            // so our timer should fire earlier
+            break;
+        }
+        prev = curr;
+        curr = curr->next;
+    }
+
+    timer->next = curr;
+    timer->fire_ticks = ticks;
+
+    if (prev != 0) {
+        prev->next = timer;
+    } else {
+        // Our timer is the first in the line to fire
+        timer_list = timer;
+        set_alarm(ticks);
+    }
+
+    // This situation might happen if adding the same timer twice
+    if (timer == timer->next) {
+        // This seems like an error: %s is used for line number
+        // In the recent SDK Espressif fixed the format to "%s %u\n"
+        printf("%s %s \n", "ets_timer.c", (char*)209);
+        while (1);
+    }
+}
+
+/**
+ * In the Espressif SDK 0.9.9 if try to arm already armed timer the system halts
+ * with error message. In the later SDK version Espressif changed the behavior.
+ * If the timer was previously armed it is disarmed and then armed without errors.
+ * This version recreates behavior of SDK 0.9.9
+ */
+void sdk_ets_timer_arm_ms_us(ets_timer_t *timer, uint32_t value,
+        bool repeat_flag, bool value_in_ms)
+{
+    uint32_t ticks = 0;
+
+    if (timer->next != ETS_TIMER_NOT_ARMED) {
+        // The error message doesn't tell what is wrong
+        printf("arm new %x %x\n", (uint32_t)timer, (uint32_t)timer->next);
+        while(1);  // halt
+    }
+
+    if (value_in_ms) {
+        value *= 1000;
+    }
+
+    if (value != 0) {
+        // Why to do multiplication for values greater than 858
+        // and do 'shift and add' for other?
+        // What is the magic number 858 ?
+        if (858 < value) {
+            ticks = (value << 2) + value;  // ticks = value * 5
+        } else {
+            // It is the same as just multiply by 5
+            // No idea why to do it this way
+            ticks = (value * 5000000) / 1000000;
+        }
+    }
+
+    if (repeat_flag) {
+        timer->period_ticks = ticks;
+    }
+    vPortEnterCritical();
+    add_pending_timer(TIMER_FRC2.COUNT + ticks, timer);
+    vPortExitCritical();
+}
+
+void sdk_ets_timer_arm(ets_timer_t *timer, uint32_t milliseconds,
```

Comment:
The sdk has the following that might as well be included. Then perhaps make `sdk_ets_timer_arm_ms_us` static.

```
void sdk_ets_timer_arm_us(ets_timer_t *timer, uint32_t useconds,
        bool repeat_flag)
{
    sdk_ets_timer_arm_ms_us(timer, useconds, repeat_flag,
            /*value in ms=*/false);
}
```